src/request_handler.cpp

```diff
@@ -16,24 +16,176 @@
 
 #include "request_handler.hpp"
 
+#include "batch_request.hpp"
 #include "connection.hpp"
 #include "constants.hpp"
+#include "error_response.hpp"
 #include "execute_request.hpp"
 #include "io_worker.hpp"
 #include "pool.hpp"
-#include "prepare_handler.hpp"
+#include "prepare_request.hpp"
+#include "response.hpp"
 #include "result_response.hpp"
 #include "row.hpp"
-#include "schema_change_handler.hpp"
+#include "schema_change_callback.hpp"
 #include "session.hpp"
 
 #include <uv.h>
 
 namespace cass {
 
-void RequestHandler::on_set(ResponseMessage* response) {
-  assert(connection_ != NULL);
-  assert(!is_query_plan_exhausted_ && "Tried to set on a non-existent host");
+class PrepareCallback : public SimpleRequestCallback {
+public:
+  PrepareCallback(const std::string& query, SpeculativeExecution* speculative_execution)
+    : SimpleRequestCallback(Request::ConstPtr(new PrepareRequest(query)))
+    , speculative_execution_(speculative_execution) { }
+
+private:
+  virtual void on_internal_set(ResponseMessage* response);
+  virtual void on_internal_error(CassError code, const std::string& message);
+  virtual void on_internal_timeout();
+
+private:
+  SpeculativeExecution::Ptr speculative_execution_;
+};
+
+void PrepareCallback::on_internal_set(ResponseMessage* response) {
+  switch (response->opcode()) {
+    case CQL_OPCODE_RESULT: {
+      ResultResponse* result =
+          static_cast<ResultResponse*>(response->response_body().get());
+      if (result->kind() == CASS_RESULT_KIND_PREPARED) {
```

Comment:
The request has been chained and we're retrying the original request. We've attempted to run a prepared query and the server responded with `UNPREPARED` so we've wrapped the original request in a prepared request and are now returning to the original request.